core/src/main/scala/kafka/server/metadata/BrokerMetadataPublisher.scala

```diff
@@ -118,6 +118,11 @@ class BrokerMetadataPublisher(conf: KafkaConfig,
    */
   var _firstPublish = true
 
+  /**
+   * This is updated after all components (e.g. LogManager) has finished publishing the new metadata delta
+   */
+  var publishedOffset: Long = -1
```

Comment:
I wonder if we should make this an AtomicLong so that it is safe to access from other threads?

Reply:
Yes, this variable is accessed in multiple threads, making it AtomicLong can avoid some problems similar to phantom read, I changed it to an AtomicLong.